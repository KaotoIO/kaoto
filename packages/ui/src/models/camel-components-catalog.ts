import { CatalogKind } from './catalog-kind';

export interface ICamelComponentDefinition {
  component: ICamelComponent;
  componentProperties: ICamelComponentProperties;
  headers: ICamelComponentHeaders;
  properties: ICamelComponentProperties;
}

export interface ICamelComponent {
  kind: CatalogKind.Component;
  name: string;
  title: string;
  description: string;
  deprecated: boolean;
  firstVersion?: string;
  label: string;
  javaType?: string;
  supportLevel?: string;
  groupId?: string;
  artifactId?: string;
  version: string;
  scheme?: string;
  extendsScheme?: string;
  syntax?: string;
  async?: boolean;
  api?: boolean;
  consumerOnly?: boolean;
  producerOnly?: boolean;
  lenientProperties?: boolean;
}

export interface CatalogItemCommon {
  index: number;
  kind: string;
  displayName: string;
  group: string;
  label: string;
  required: boolean;
  javaType: string;
  deprecated: boolean;
  deprecationNote: string;
  autowired: boolean;
  secret: boolean;
  description: string;
}

export interface ICamelComponentProperties {
  type: string;
}

export interface ICamelComponentHeaders extends CatalogItemCommon {
  constantName: string;
}
