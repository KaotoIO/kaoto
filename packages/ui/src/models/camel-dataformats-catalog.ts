import { ICamelProcessorModel, ICamelProcessorProperty } from './camel-processors-catalog';
import { KaotoSchemaDefinition } from './kaoto-schema';

export interface ICamelDataformatDefinition {
  model: ICamelDataformatModel;
  properties: Record<string, ICamelDataformatProperty>;
  propertiesSchema: KaotoSchemaDefinition['schema'];
}

export type ICamelDataformatModel = ICamelProcessorModel;

export type ICamelDataformatProperty = ICamelProcessorProperty;
