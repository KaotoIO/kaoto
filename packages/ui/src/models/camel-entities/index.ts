export * from './base-entity';

/** Camel Route definitions */
export * from './flows/route';
