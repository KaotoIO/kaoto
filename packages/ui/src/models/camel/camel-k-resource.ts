import {
  Integration as IntegrationType,
  KameletBinding as KameletBindingType,
  Pipe as PipeType,
} from '@kaoto/camel-catalog/types';
import { TileFilter } from '../../components/Catalog';
import { createCamelPropertiesSorter } from '../../utils';
import { IKameletDefinition } from '../kamelets-catalog';
import { AddStepMode, BaseVisualCamelEntity, IVisualizationNodeData } from '../visualization/base-visual-entity';
import { MetadataEntity } from '../visualization/metadata';
import { BaseVisualCamelEntityDefinition, CamelResource } from './camel-resource';
import { BaseCamelEntity } from './entities';
import { SourceSchemaType } from './source-schema-type';
import { CamelResourceSerializer, YamlCamelResourceSerializer } from '../../serializers';

export type CamelKType = IntegrationType | IKameletDefinition | KameletBindingType | PipeType;

export enum CamelKResourceKinds {
  Integration = 'Integration',
  Kamelet = 'Kamelet',
  KameletBinding = 'KameletBinding',
  Pipe = 'Pipe',
}

export const CAMEL_K_K8S_API_VERSION_V1 = 'camel.apache.org/v1';

export abstract class CamelKResource implements CamelResource {
  static readonly PARAMETERS_ORDER = ['apiVersion', 'kind', 'metadata', 'spec', 'source', 'steps', 'sink'];
  // static serializer = new YamlResourceSerializer();
  readonly sortFn = createCamelPropertiesSorter(CamelKResource.PARAMETERS_ORDER) as (a: unknown, b: unknown) => number;
  protected resource: CamelKType;
  private metadata?: MetadataEntity;

  constructor(
    parsedResource: unknown,
    private readonly serializer: CamelResourceSerializer = new YamlCamelResourceSerializer(),
  ) {
    this.serializer = serializer;

    if (parsedResource) {
      this.resource = parsedResource as CamelKType;
    } else {
      this.resource = {
        apiVersion: CAMEL_K_K8S_API_VERSION_V1,
        spec: {},
      };
    }
    this.metadata = this.resource.metadata && new MetadataEntity(this.resource);
  }

  getCanvasEntityList(): BaseVisualCamelEntityDefinition {
    return {
      common: [],
      groups: {},
    };
  }

  removeEntity(_id?: string) {}
  refreshVisualMetadata() {}
  createMetadataEntity() {
    this.resource.metadata = {};
    this.metadata = new MetadataEntity(this.resource);
    return this.metadata;
  }

  getMetadataEntity() {
    return this.metadata;
  }

  deleteMetadataEntity() {
    this.resource.metadata = undefined;
    this.metadata = undefined;
  }

  getEntities(): BaseCamelEntity[] {
    const answer = [];
    if (this.resource.metadata && this.metadata) {
      answer.push(this.metadata);
    }
    return answer;
  }

  abstract getType(): SourceSchemaType;

  abstract getVisualEntities(): BaseVisualCamelEntity[];

  abstract toJSON(): unknown;

  addNewEntity(): string {
    /** Not supported by default */
    return '';
  }

  supportsMultipleVisualEntities(): boolean {
    /** Not supported by default */
    return false;
  }

  /** Components Catalog related methods */
  getCompatibleComponents(_mode: AddStepMode, _visualEntityData: IVisualizationNodeData): TileFilter | undefined {
    return undefined;
  }
  getSerializer() {
    return this.serializer;
  }

  setSerializer(_serializer: CamelResourceSerializer): void {
    /** Not supported by default */
  }

  toString(): string {
    return this.serializer.serialize(this);
  }
}
