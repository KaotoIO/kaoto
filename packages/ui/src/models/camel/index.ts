export * from './camel-resource';
export * from './camel-route-resource';
export * from './integration-resource';
export * from './kamelet-resource';
export * from './kamelet-binding-resource';
export * from './pipe-resource';
export * from './source-schema-config';
export * from './source-schema-type';
export { CAMEL_K_K8S_API_VERSION_V1 } from './camel-k-resource';
