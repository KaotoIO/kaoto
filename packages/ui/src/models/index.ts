export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
