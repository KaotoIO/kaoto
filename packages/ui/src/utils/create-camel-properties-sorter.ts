import { Pair } from 'yaml';

export type PairKey = { value: string };
/**
 * Sort properties in a Apache Camel friendly format.
 *
 * The order goes as follows:
 *   1. `id` prop
 *   2. `description` prop
 *   3. `uri` prop
 *   4. `parameters` prop
 *   5. `steps` prop
 *   6. the remaining properties in alphabetical order
 */
export const createCamelPropertiesSorter =
  (order: string[]) => (a: Pair<PairKey, unknown>, b: Pair<PairKey, unknown>) => {
    const aIndex = order.indexOf(a.key.value);
    const bIndex = order.indexOf(b.key.value);

    if (aIndex === -1 && bIndex === -1) {
      return a.key.value.localeCompare(b.key.value);
    }

    if (aIndex === -1) {
      return 1;
    }

    if (bIndex === -1) {
      return -1;
    }

    return aIndex - bIndex;
  };
