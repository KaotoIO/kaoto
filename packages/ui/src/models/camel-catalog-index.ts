import { ICamelComponentDefinition } from './camel-components-catalog';
import { ICamelProcessorDefinition } from './camel-processors-catalog';
import { CatalogKind } from './catalog-kind';
import { IKameletDefinition } from './kamelets-catalog';

export interface CamelCatalogIndex {
  catalogs: Catalogs;
  schemas: CatalogEntry[];
}

export interface Catalogs {
  models: CatalogEntry;
  components: CatalogEntry;
  languages: CatalogEntry;
  dataformats: CatalogEntry;
  kamelets: CatalogEntry;
}

export interface CatalogEntry {
  name: string;
  version: string;
  file: string;
}

export type CatalogTypes = Record<string, ICamelComponentDefinition | ICamelProcessorDefinition | IKameletDefinition>;

export interface ComponentsCatalog {
  [CatalogKind.Component]?: Record<string, ICamelComponentDefinition>;
  [CatalogKind.Processor]?: Record<string, ICamelProcessorDefinition>;
  [CatalogKind.Kamelet]?: Record<string, IKameletDefinition>;
  [key: string]: unknown;
}

export interface CatalogCamelComponent {
  type: CatalogKind.Component;
  definition: ICamelComponentDefinition;
}

export interface CatalogCamelProcessor {
  type: CatalogKind.Processor;
  definition: ICamelProcessorDefinition;
}

export interface CatalogKamelet {
  type: CatalogKind.Kamelet;
  definition: IKameletDefinition;
}
