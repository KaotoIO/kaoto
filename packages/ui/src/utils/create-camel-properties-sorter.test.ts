import { Pair, parse, stringify } from 'yaml';
import { CamelRouteResource } from '../models/camel';
import { CamelKResource } from '../models/camel/camel-k-resource';
import { createCamelPropertiesSorter } from './create-camel-properties-sorter';

describe('camelPropertiesSorter', () => {
  it('should sort the Camel Route', () => {
    const camelRouteYaml = `
- route:
    id: route-1398
    from:
      description: my descriptio
      uri: timer
      id: "1"
      steps:
        - to:
            parameters:
              disableReplyTo: true
              synchronous: true
              transferException: true
            uri: amqp
            id: to-8385
        - to:
            id: to-2786
            uri: whatsapp
            parameters:
              baseUri: /
              lazyStartProducer: true
      parameters:
        delay: "{{delay}}"
        fixedRate: true
        includeMetadata: true
        period: 50.75
        synchronous: true
        timerName: timer-1-1`;

    const rawObject = parse(camelRouteYaml);

    const result = stringify(rawObject, {
      sortMapEntries: createCamelPropertiesSorter(CamelRouteResource.PARAMETERS_ORDER) as (a: Pair, b: Pair) => number,
    });

    expect(result).toMatchSnapshot();
  });

  it('should sort Pipes', () => {
    const pipeYaml = `
    spec:
      steps:
        - ref:
            apiVersion: camel.apache.org/v1
            kind: Kamelet
            name: delay-action
            properties:
              milliseconds: "100"
      sink:
        ref:
          apiVersion: camel.apache.org/v1
          kind: Kamelet
          name: log-sink
      source:
        ref:
          apiVersion: camel.apache.org/v1
          kind: Kamelet
          name: timer-source
          properties:
            message: hello
            repeatCount: jjj
    kind: Pipe
    metadata:
      name: pipe-1748
    apiVersion: camel.apache.org/v1`;

    const rawObject = parse(pipeYaml);
    const result = stringify(rawObject, {
      sortMapEntries: createCamelPropertiesSorter(CamelKResource.PARAMETERS_ORDER) as (a: Pair, b: Pair) => number,
    });

    expect(result).toMatchSnapshot();
  });
});
