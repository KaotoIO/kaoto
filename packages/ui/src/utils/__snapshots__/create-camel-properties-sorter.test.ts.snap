// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`camelPropertiesSorter should sort Pipes 1`] = `
"apiVersion: camel.apache.org/v1
kind: Pipe
metadata:
  name: pipe-1748
spec:
  source:
    ref:
      apiVersion: camel.apache.org/v1
      kind: Kamelet
      name: timer-source
      properties:
        message: hello
        repeatCount: jjj
  steps:
    - ref:
        apiVersion: camel.apache.org/v1
        kind: Kamelet
        name: delay-action
        properties:
          milliseconds: "100"
  sink:
    ref:
      apiVersion: camel.apache.org/v1
      kind: Kamelet
      name: log-sink
"
`;

exports[`camelPropertiesSorter should sort the Camel Route 1`] = `
"- route:
    id: route-1398
    from:
      id: "1"
      description: my descriptio
      uri: timer
      parameters:
        delay: "{{delay}}"
        fixedRate: true
        includeMetadata: true
        period: 50.75
        synchronous: true
        timerName: timer-1-1
      steps:
        - to:
            id: to-8385
            uri: amqp
            parameters:
              disableReplyTo: true
              synchronous: true
              transferException: true
        - to:
            id: to-2786
            uri: whatsapp
            parameters:
              baseUri: /
              lazyStartProducer: true
"
`;
